src/boot/bootctl-status.c

```diff
@@ -503,6 +504,224 @@ int verb_status(int argc, char *argv[], void *userdata) {
+                        if (r < 0)
+                                return r;
+                }
+        }
+
+        *ret_known_files = TAKE_PTR(known_files);
+
+        return 0;
+}
+
+static int unlink_entry(const BootConfig *config, const char *root, const char *fn) {
+        _cleanup_(hashmap_free_free_keyp) Hashmap *known_files = NULL;
+        const BootEntry *e = NULL;
+        int r;
+
+        r = count_known_files(config, root, &known_files);
+        if (r < 0)
+                return log_error_errno(r, "Failed to count files in %s: %m", root);
+
+        for (size_t i = 0; i < config->n_entries; i++) {
+                const BootEntry *ei = &config->entries[i];
+
+                if (!ei->root || strcmp(ei->root, root))
+                        continue;
+                if (fnmatch(fn, ei->id, FNM_CASEFOLD))
+                        continue;
+
+                e = ei;
+                if (i == (size_t) config->default_entry)
+                        log_warning("%s is the default boot entry", fn);
+                if (i == (size_t) config->selected_entry)
+                        log_warning("%s is the selected boot entry", fn);
+                break;
+        }
+
+        if (!e)
+                return -ENOENT;
+
+        if (e->kernel) {
+                _cleanup_free_ char *d = NULL;
+
+                deref_unlink_file(known_files, e->kernel, e->root);
+                STRV_FOREACH(s, e->initrd)
+                        deref_unlink_file(known_files, *s, e->root);
+                deref_unlink_file(known_files, e->device_tree, e->root);
+                STRV_FOREACH(s, e->device_tree_overlay)
+                        deref_unlink_file(known_files, *s, e->root);
+
+                if (path_extract_directory(e->kernel, &d) == 0)
+                        (void) rmdir(d);
+        }
+
+        if (!arg_dry) {
+                r = unlink(e->path);
+                if (r < 0)
+                        return log_error_errno(errno, "failed to remove \"%s\": %m", e->path);
+        }
+
+        log_info("%s %s", arg_dry?"Would remove":"Removed", e->path);
+
+        return 0;
+}
+
+struct list_remove_orphaned_file_data  {
+        const char *root;
+        Hashmap *known_files;
+};
+
+static int list_remove_orphaned_file(
+                RecurseDirEvent event,
+                const char *path,
+                int dir_fd,
+                int inode_fd,
+                const struct dirent *de,
+                const struct statx *sx,
+                void *userdata) {
+        struct list_remove_orphaned_file_data *data = userdata;
+        int r;
+
+        assert(path);
+        assert(data);
+        assert(data->known_files);
+        assert(data->root);
+
+        /* shouldn't happen */
+        if (!startswith(path, data->root))
+                return RECURSE_DIR_LEAVE_DIRECTORY;
+
+        if (event == RECURSE_DIR_ENTRY) {
+                if (hashmap_get(data->known_files, path+strlen(data->root)) == NULL) {
```

Comment:
it's always better to prefix a path with the root than to chop it off again.

hece, instead of doing `path+strlen(data->root)` here, just call recurse_dir() with a NULL as second argument. that way you will be called here with `path` just being the path relative to the passed in fd. which is what you actually want.

then, if you want to prefix the root, do so manually below for the logging.

Reply:
Ah, I didn't know that trick with recurse_dir(). I don't mind but adding the prefix would require path_join ie alloc, strcpy and free in every step just for logging.

Reply:
... or I just add one `log_info("cleaning %s", root);` before `recurse_dir` :thinking: